text/0006-wasm-smart-contracts.md

```diff
@@ -0,0 +1,440 @@
+    world.create(State.from(new_asset));
+```
+
+## Transactional Information
+
+### Current Transaction 
+
+To obtain the current transaction and the essential information:
+
+```rust
+    let transaction = Transaction::current_transaction();
+    transaction.get_timestamp();
+    transaction.get_id();
+    transaction.get_mspid();
+```
+
+`get_mspid` is important for handling situations involving private data and being able to determine the originating organization. 
+
+### Events
+
+Currently, a single event payload and name can be produced from a chaincode. Feedback has indicated that at a minimum it should be possible to have multiple names and payloads.
+
+The transaction, therefore, can support this
+
+```rust
+    transaction.set_event(evtName, eventData);
+    transaction.get_event(evtName, eventData);
+```
+
+These will then be put into a JSON Object structure and returned. The further specification will be left to later work. 
+
+### Identity
+
+Information about the identity that requested the transaction can be obtained:
+
+```rust
+        let identity = transaction.get_client_identity();
+        identity.attribute_exists(attrName,attrValue);
+        identity.attribute_value(attrName);
+        identity.getID()
+        identity.getMSPID();
+```
+
+Will validate that this is the appropriate set of the accessors to support against real use-cases. 
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Wasm Runtime
+
+These are the major components within the runtime.
+
+- a Wasm runtime engine responsible for the execution of the Wasm binary
+- grpc communications with the peer
+- data marshalling for transfer of data and requests into and out of the Wasm engine
+- Along with required code to bootstrap the whole Wasm engine and load in the binary
+
+### Wasm Engine
+Several different implementations are available; initial prototypes have used the LifeVM, but for final implementation, Wasmer is being considered. The WebAssembly landscape is still developing so there is some risk with changing standards and implementations, however, some of the new standards could be of benefit in the future.
+
+### grpc Peer Communications
+This will use the go chaincode shim today to provide connectivity to the peer.
+
+### Data Marshalling
+Currently, the API to import/export functions and data between the host and the executing code in the wasm runtime is limited to primitive data types only. A data and function marshalling interface is required, waPC will be used to provide this. 
+
+The logical functions that will be marshalled over waPC from host to client will form part of the deliverables. This is to allow for new languages to be supported and also to permit changes to the peer communications protocol. (these logical functions are referred to as the Contract Core SPI). It will be designed to use the protobuf specification to permit accurate description.
```

Comment:
Hi! 👋 I am one of the creators of waPC.  I think it is so awesome that you are considering it for smart contracts.  As you've experienced, waPC only handles sending payloads as byte arrays and leaves it to the implementation to handle serialization/deserialization.

I wanted to briefly share our findings using Protobuf.  In Rust, you will find Protobuf works great (everything for the most part in Rust compiles down to Wasm as expected).  In languages like TinyGo (The best Wasm compiler for Go outside the browser right now), reflection and some basic standard libraries that Protos need are not implemented.  Additionally, we really like [AssemblyScript](https://docs.assemblyscript.org/) but likewise there is no Protobuf support.

Because of this, we decided to use MessagePack as the serialization format over Protobuf.  We even started experimenting with a "fit-for-purpose" waPC IDL (called WIDL) and [code generator](https://github.com/wapc/as-codegen) that creates all the necessary glue code (waPC + MsgPack).  So far, developers seem to like the simplicity.  Here is an [example](https://github.com/wapc/as-actor-test).  We'll be releasing an improved code generator that supports extensibility soon.

We hoped that these tools would be needed only for a short time until [interface types](https://github.com/WebAssembly/interface-types) arrived.  Unfortunately, (and I could be wrong) it does not seem like interface types solve for exactly what waPC is trying to accomplish.  It is still early days and I'm hopeful it will take shape.  Regardless, feel free to reach out if you need anything regarding waPC.  Cheers!

Reply:
Thanks for your comment; and the information about MessagePack vs Protobuf. We already have protobuf in use in the peer-chaincode coms protocols so didn't ideally want to add another piece of technology. 

When I looked at Interface Types initially I likewise wasn't quite sure; waPC makes a pretty good fit for what I wanted.  And we don't need to write our own go host now... though do I have a node.js version for testing.

Reply:
Yeah, great point @mbwhite.  After posting this I got curious about creating a Protobuf library for AssemblyScript and found that at least the library and hand-coding marshaling/unmarshaling would be relatively easy.  It might even take less coding time than our [MsgPack](https://github.com/wapc/as-msgpack) library.  The harder part might be creating a `protoc` plugin.  I'll try to find time to investigate that but it all feels very doable.  My feeling is that AssemblyScript, being so similar to TypeScript/JavaScript, might be more the "target audience" vs. Rust developers but obviously supporting both is ideal.

Reply:
For some later work, I've prototyped a protoc plugin;  the basic structure looks confusing at first but I was lucky to find a good example;  the actual code generation part.. well that's a different matter!

Reply:
Regarding protobuf for AssemblyScript. Whay not use Prot3 bindings for JSON and manipulate the proto message entirely in json format in AssemblyScript ( we are already using protobuf serialized json objects inside AssemblyScript this way )